### ISSUE TITLE
App titlebar does not follow system theme Fedora35/Gnome/Wayland

### ISSUE THREAD

**Issue by AI-ien** [polychromatic/polychromatic#400]
Displayed titlebar is always 'default' no matter what theme is currently active. Toggling 'Use system theme' in preferences does the same, it picks 'default' not active theme.
**app version: 0.7.3**

![image](https://user-images.githubusercontent.com/33252703/148537308-6804b6b6-7c7e-49c4-80b0-075a39023937.png)

### Operating System / Desktop Environment / Version(s)
![image](https://user-images.githubusercontent.com/33252703/148537615-aa0dea33-c683-47d1-8eca-b33c6bd67795.png)

**Comment by lah7** [comment]
Do you mean the title bar should be dark being a dark-themed app? If so, this would be the same as #352. In that case, there isn't a reliable way to fix this, since Qt doesn't have an API to provide "use dark" hints. I'd love to see that fixed.

Or do you mean the title bar looks completely different to other applications (say, you use Arc theme but this app shows a Adwaita titlebar). If that's the case, this might indicate a problem with how Fedora integrates Qt applications.

**Comment by AI-ien** [comment]
Yes, I mean the second. Because it is not just about title bar, toggling 'use system theme' will also load default adwaita instead of current.

**Comment by lah7** [comment]
It seems that Fedora 35 doesn't integrate Qt and GTK applications equally by default. It's up to distros to provide a configuration so Qt apps (like this one) can support the native look on GTK/GNOME desktops. Ubuntu MATE and Linux Mint are good examples.

According to [this reddit post](https://www.reddit.com/r/Fedora/comments/cs1u8i/how_to_make_qt_applications_look_like_the_rest_of/exe8ev3), this might work:
```bash
dnf install qt5-qtstyleplugins 
echo "export QT_QPA_PLATFORMTHEME=gtk2" >> ~/.profile
```
_(Log out and back in to take effect)_

If this doesn't, [the Arch Wiki may be useful](https://wiki.archlinux.org/title/Uniform_look_for_Qt_and_GTK_applications). I'm unfamiliar with Fedora, so their forums would be better outlet for advice, especially with Wayland. If there's any open bug reports about Qt theme integration by default, you may wish to give those a +1. Other Qt-based applications like VLC will be affected by this too.

**Comment by AI-ien** [comment]
Your suggestion did work, but it made all look like Windows 98.
Nevertheless there are few options (none that solves header bar theming so). Installing qt5-qtstyleplugins you get available styles: HighContrastInverse, HighContrast, Adwaita-HighContrastInverse, Adwaita-HighContrast, Adwaita-Dark, Adwaita, bb10dark, bb10bright, kvantum-dark, kvantum, cleanlooks, gtk2, cde, motif, plastique, Windows, Fusion. Those can be exported with `export QT_STYLE_OVERRIDE=kvantum` for example.

Another option I found is to install kvantum-manager that allows to use more themes.
Going off-topic. What about packing as flatpak? I think this way you get better compatibility, system headerbars with Qt widgets inside app.

**Comment by lah7** [comment]
I know Linux Mint and Ubuntu MATE use `gtk2` engine for the Qt style, and that pulls in styling from the GTK (2?) theme. I think they include an additional package to provide this theme engine.

>  (none that solves header bar theming so)

Is the title bar completely different to other applications? Please could you show me a screenshot (and name of theme)?

> Going off-topic. What about packing as flatpak?

I haven't looked too much into it (#394) since the app is _natively_ packaged for the major distros, so it's at the bottom of the priority list. My main concern is that Flatpak/Snap promote a sandbox environment so I'm not sure how this would work when communicating with OpenRazer (and others, in future) and use external libraries.

**Comment by AI-ien** [comment]
I use [Materia](https://github.com/nana-4/materia-theme) theme by nana-4, plus I use `~/.local/share/gtk-3.0/gtk.css` to override titlebar icons from [here](https://github.com/AI-ien/macos-titlebar), but that all is not important, since it will happen with any theme that is not Adwaita. 

If I'm not wrong app is entirely built using Qt, including main window. Thus Fedora uses some compatibility layer of qt5 library and applies one of ported themes to draw window. If system theme does not matches it uses default Adwaita window decorations. 

Ideally app should be built to run using host window system thus using local window theme, something similar how electron apps work.

**VLC (flatpak) titlebar**
![Screenshot from 2022-01-09 22-59-33](https://user-images.githubusercontent.com/33252703/148701727-df39bf74-7a85-4fb3-9689-b532012d7c40.png)

**Comment by lah7** [comment]
I'm pretty sure it's down to Wayland, the distro and/or the window system to handle the integration of _"second class"_ toolkits like Qt in a GNOME/GTK environment. Electron uses GTK so those apps are not affected.

I don't understand [how/if VLC/Flatpak does it](https://github.com/flathub/org.videolan.VLC), but I'm guessing something, somewhere performs "magic" so it integrates just fine for VLC/Flatpak. Another PyQt5 app I know is `git-cola`, I bet the same problem happens there under native conditions. I don't believe there's anything we can change in the app's code to "fix" an external integration problem, sadly.

If you wish to test other Qt-based apps, there's `git-cola` (PyQt5),  `virtualbox` (Qt 5) and KDE apps like `kwrite` (Qt 5).

According to the [Arch Wiki](https://wiki.archlinux.org/title/Uniform_look_for_Qt_and_GTK_applications), there is this:

> **Qt apps run on GNOME Wayland have a non-matching window decoration look, even after setting a Qt theme**
>
> In order to have a matching window decoration look, you have to install `qgnomeplatform`, and set the following environment variable: `QT_QPA_PLATFORMTHEME='gnome'`. This fix is guaranteed to work with Adwaita or Adwaita-dark. 

Maybe try this under Fedora? Unless of course, such integration is hardcoded to Adwaita. This wouldn't surprise me, seeing as GNOME gives the impression they don't support third party theming.

**Comment by AI-ien** [comment]
> Another PyQt5 app I know is git-cola, I bet the same problem happens there under native conditions. I don't believe there's anything we can change in the app's code to "fix" an external integration problem, sadly.
I did tried git cola and titlebar actually is displayed correctly as it should be, the same as VLC. Plus it was rpm pack from Fedora repo. 
KWrite on another hand didn't follow system theme and both rpm and flatpak packages had Adwaita titlebars.
Go figure now.

**Comment by lah7** [comment]
I tried replicating the issue on Fedora 35 with GNOME 41 under Wayland.

1. `sudo dnf install materia-gtk-theme gnome-tweaks`
2. Use GNOME Tweaks to change the theme to "Materia-dark-compact"

![image](https://user-images.githubusercontent.com/13032135/149094534-b50b8288-8951-4932-a0f2-1860739ba998.png)
![image](https://user-images.githubusercontent.com/13032135/149094545-692f9df5-fa30-4594-bb3b-e9800d255ee9.png)

Then after installing `qgnomeplatform` and setting `QT_QPA_PLATFORMTHEME='gnome'` - the control's are now themed to Adwaita-Dark, including the title bar - but it's not Materia:

![Screenshot_20220112_085419](https://user-images.githubusercontent.com/13032135/149095227-cedb7806-a330-45fe-a940-88c3932d8df1.png)

Strangely, `git-cola` (which uses the same `python3-pyqt5` package) has the correct title bar, but with Adwaita-Dark controls/colours:

![Screenshot_20220112_091810](https://user-images.githubusercontent.com/13032135/149099236-0cfa27cb-b431-4bd3-945d-d6bda7ee8179.png)

Uninstalling `qgnomeplatform` restores the title bar, but we're back to stock Qt controls (unless you use Polychromatic's Qt theme).

Even a minimal PyQt5 application does not have the correct title bar under `qgnomeplatform`: [testcase-svg-icon-states.py.gz](https://github.com/polychromatic/polychromatic/files/7853195/testcase-svg-icon-states.py.gz)

![Screenshot_20220112_091150](https://user-images.githubusercontent.com/13032135/149098263-d0dea753-a2db-4197-8a0b-6894aff7dd7b.png)

Perhaps other distributions may be affected by this, but I'd suggest raising this issue with [FedoraQt/QGnomePlatform](https://github.com/FedoraQt/QGnomePlatform) and use the minimal Python script above as an example (which I re-used from an unrelated bug report I posted for KDE months ago)

----

Unsetting `QT_QPA_PLATFORM_THEME` doesn't seem sufficient, so the workaround could be to uninstall `qgnomeplatform` **but** that will probably break the theming for VLC, and other Qt applications.
 
![Screenshot_20220112_092728](https://user-images.githubusercontent.com/13032135/149103750-4ac493c0-3408-450c-933c-80d7f3d4c0b4.png)
![Screenshot_20220112_093156](https://user-images.githubusercontent.com/13032135/149106314-cd50c6e1-e21c-4056-8d61-ee5b9e3cbbfd.png)

The title bar is not a problem under X11, but the controls are still Adwaita-Dark.

![Screenshot_20220112_093803](https://user-images.githubusercontent.com/13032135/149109840-82959d71-9512-42df-8752-24235f655236.png)
